evm/rpc/server.py

```diff
@@ -0,0 +1,176 @@
+"""
+A simple JSON-RPC server that only responds to eth_getBlockByNumber and eth_getBlockByHash
+calls. It uses the LightChain to sync headers as they're announced and fetches blocks on demand
+as RPC calls ask for them.
+"""
+import asyncio
+import logging
+
+import rlp
+
+from aiohttp import web
+from aiohttp.web_exceptions import HTTPMethodNotAllowed
+
+from eth_utils import decode_hex, encode_hex
+
+from eth_keys import keys
+
+from evm.chains.mainnet import MAINNET_VM_CONFIGURATION
+from evm.chains.ropsten import (
+    ROPSTEN_GENESIS_HEADER,
+    ROPSTEN_NETWORK_ID,
+)
+from evm.p2p import ecies
+from evm.p2p import kademlia
+from evm.p2p.constants import HANDSHAKE_TIMEOUT
+from evm.p2p.lightchain import (
+    LightChain,
+    OnDemandDataBackend,
+)
+from evm.p2p.peer import (
+    handshake,
+    LESPeer,
+)
+from evm.utils.numeric import int_to_big_endian
+
+
+# Change the values below to connect to a node on a different network or IP address.
+GENESIS_HEADER = ROPSTEN_GENESIS_HEADER
+NETWORK_ID = ROPSTEN_NETWORK_ID
+# The pubkey for the local node we'll connect to. Simply pass the hex string below to
+# geth using the "-nodekeyhex" argument.
+NODE_ID = keys.PrivateKey(decode_hex(
+    "45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8")).public_key
+NODE_ADDR = kademlia.Address('127.0.0.1', 30303, 30303)
+
+
+class App(web.Application):
+    allowed_methods = ['eth_getBlockByNumber', 'eth_getBlockByHash']
+
+    def __init__(self, chain):
+        super(App, self).__init__()
+        self.chain = chain
+        self.router.add_post('/', self.handle)
+        self.on_startup.append(self.connect_peer)
+        self.on_shutdown.append(self.stop_chain)
+
+    @asyncio.coroutine
+    def handle(self, request):
+        body = yield from request.json()
+        req_id = body['id']
+        method = body['method']
+        hash_or_number, _ = body['params']
+        if method == 'eth_getBlockByNumber':
+            if hash_or_number == "latest":
+                head = self.chain.get_canonical_head()
+                number = head.block_number
+            else:
+                number = int(hash_or_number, 16)
+            block = yield from self.chain.get_canonical_block_by_number(number)
+        elif method == 'eth_getBlockByHash':
+            block_hash = decode_hex(hash_or_number)
+            block = yield from self.chain.get_block_by_hash(block_hash)
+        else:
+            raise HTTPMethodNotAllowed(method, self.allowed_methods)
```

Comment:
Maybe just add a `TODO` here that instead of raising an exception we should instead return an error response over the RPC server.

Reply:
The exception gets caught by aiohttp and turned into a 405 response. Ultimately, in web3.py we get a `HTTPError: 405 Client Error: Method Not Allowed for url: http://localhost:8080/`